src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramBuilder.cpp

```diff
@@ -696,11 +696,59 @@ class ShaderCallDither : public ShaderPart
-				"  if (uAlphaDitherMode == 2) alphaNoiseDither(snoise(), clampedColor.a);	\n"
+		if (!_glinfo.isGLES2) {
+			if (config.generalEmulation.ditheringMode >= 1) {
+				m_part =
+					"  if (uColorDitherMode == 2) {												\n"
+					"    colorNoiseDither(snoiseRGB(), clampedColor.rgb);						\n"
+					"    quantizeRGB(clampedColor.rgb);											\n"
+					"  }																		\n"
+					"  if (uAlphaDitherMode == 2) {												\n"
+					"    alphaNoiseDither(snoiseA(), clampedColor.a);							\n"
+					"    quantizeA(clampedColor.a);												\n"
+					"  }																		\n"
+				;
+			}
+			if (config.generalEmulation.ditheringMode >= 3) {
+				m_part +=
+					// Try to keep dithering visible even at higher resolutions
+					"  lowp float divider = 1.0 + step(3.0, uScreenScale.x);					\n"
+					"  mediump ivec2 position = ivec2(mod((gl_FragCoord.xy - 0.5) / divider,4.0));\n"
+					"  																			\n"
+					"  lowp mat4 bayer = mat4( 0.0, 0.75, 0.1875, 0.9375,						\n"
+					"                            0.5, 0.25, 0.6875, 0.4375,						\n"
+					"                            0.125, 0.875, 0.0625, 0.8125,					\n"
+					"                            0.625, 0.375, 0.5625, 0.3125);					\n"
+					"  lowp mat4 mSquare = mat4( 0.0, 0.6875, 0.75, 0.4375,						\n"
+					"                              0.875, 0.3125, 0.125, 0.5625, 				\n"
+					"                              0.1875, 0.5, 0.9375, 0.25,					\n"
+					"                              0.8125, 0.375, 0.0625, 0.625);				\n"
+					"  																			\n"
+					"  lowp float bayerThreshold = bayer[position.x][position.y];				\n"
+					"  lowp float mSquareThreshold = mSquare[position.x][position.y];			\n"
+					"  																			\n"
+					"  if (uColorDitherMode < 2) {												\n"
+					"    lowp float threshold = mix(mSquareThreshold,bayerThreshold,float(uColorDitherMode));\n"
+					"    colorNoiseDither(vec3(threshold), clampedColor.rgb);					\n"
+						// 16 Bit quantization
+					"    quantizeRGB(clampedColor.rgb);											\n"
+					"  }																		\n"
+					"  if (uAlphaDitherMode < 2) {												\n"
+					"    lowp float thresholdPattern = mix(mSquareThreshold,bayerThreshold,clamp(float(uColorDitherMode),0.0,1.0));\n"
+					"    thresholdPattern = mix(thresholdPattern,mSquareThreshold,clamp(float(uColorDitherMode - 1),0.0,1.0));\n" // uColorDitherMode = 2
+					"    thresholdPattern = mix(thresholdPattern,bayerThreshold,clamp(float(uColorDitherMode - 2),0.0,1.0));\n" // uColorDitherMode = 3
+					"  																			\n"
+					"    lowp float thresholdNotPattern = mix(bayerThreshold,mSquareThreshold,clamp(float(uColorDitherMode),0.0,1.0));\n"
+					"    thresholdNotPattern = mix(thresholdNotPattern,bayerThreshold,clamp(float(uColorDitherMode - 1),0.0,1.0));\n" // uColorDitherMode = 2
+					"    thresholdNotPattern = mix(thresholdNotPattern,mSquareThreshold,clamp(float(uColorDitherMode - 2),0.0,1.0));\n" // uColorDitherMode = 3
```

Comment:
Why some many calculations for alpha dither?
Why alpha dither depends on uColorDitherMode?

Reply:
I just wanted to replace branches with mix functions. Should not be necessary for newer gpus. 

Alpha dither depends on uColorDitherMode if alpha dither mode is G_AD_PATTERN or G_AD_NOTPATTERN according to this docu:

https://hack64.net/wiki/doku.php?id=rcpstructs

Reply:
> I just wanted to replace branches with mix functions. Should not be necessary for newer gpus.

Since uColorDitherMode is an uniform, we can use it in if..else statement without noticeable performance penalty. We can even use switch(uColorDitherMode) since you disabled shader dithering for GLES2.

 >Alpha dither depends on uColorDitherMode if alpha dither mode is G_AD_PATTERN or G_AD_NOTPATTERN according to this docu:

Nice docu, thanks!